src/metabase/models/field_usage.clj

```diff
@@ -0,0 +1,104 @@
+   :filter_op              mi/transform-keyword
+   :filter_args            mi/transform-json})
+
+(t2/define-before-insert :model/FieldUsage
+  [instance]
+  (merge {:timestamp :%now}
+         instance))
+
+(defn- filter->field-usage
+  [query stage filter-clause]
+  (let [filter-parts (lib/filter-parts query stage filter-clause)
+        field-id     (-> filter-parts :column :id)]
+    (when (int? field-id)
+      {:field_id    field-id
+       :used_in     :filter
+       :filter_op   (-> filter-parts :operator :short)
+       :filter_args (:args filter-parts)})))
+
+(defn- aggregation->field-usage
+  [query stage aggregation-clause]
+  (let [aggregation-column (lib/aggregation-column query stage aggregation-clause)
+        field-id           (:id aggregation-column)]
+    (when (int? field-id)
+      {:field_id             field-id
+       :used_in              :aggregation
+       :aggregation_function (first aggregation-clause)})))
+
+(defn- breakout->field-usage
+  [query stage breakout-clause]
+  (let [breakout-column (lib/breakout-column query stage breakout-clause)
+        field-id        (:id breakout-column)]
+    (when (int? field-id)
+      {:field_id               field-id
+       :used_in                :breakout
+       :breakout_temporal_unit (lib/raw-temporal-bucket breakout-clause)
+       :breakout_binning       (-> (lib/binning breakout-clause)
+                                   (select-keys [:strategy :bin-width :num-bins])
+                                   not-empty)})))
+
+(defn- expression->field-usage
+  [expresison-clause]
+  (when-let [field-ids (seq (lib.util.match/match expresison-clause [:field _ (id :guard int?)] id))]
+    (for [field-id field-ids]
+      {:field_id field-id
+       :used_in  :expression})))
+
+(defmulti ^:private pquery->field-usages-method
+  "Impl for [[pquery->field-usages]]."
+  {:arglists '([query stage-number x])}
+  (fn [_query _stage-number x] (lib.dispatch/dispatch-value x))
+  :hierarchy lib.hierarchy/hierarchy)
+
+(defmethod pquery->field-usages-method :mbql/query
+  [query _stage-number x]
+  (flatten (for [stage (range (lib/stage-count x))]
+                (pquery->field-usages-method query stage (lib.util/query-stage query stage)))))
+
+(defmethod pquery->field-usages-method :metabase.lib.stage/stage
+  [query stage-number x]
+  (concat
+    (keep #(filter->field-usage query stage-number %) (lib/filters query stage-number))
+    (keep #(aggregation->field-usage query stage-number %) (lib/aggregations query stage-number))
+    (keep #(breakout->field-usage query stage-number %) (lib/breakouts query stage-number))
+    (flatten (keep expression->field-usage (lib/expressions query stage-number)))
+    (flatten (keep #(pquery->field-usages-method query stage-number %) (:joins x)))))
+
+(defmethod pquery->field-usages-method :mbql/join
+  [query _stage-number x]
+  (let [join-query (fetch-source-query/resolve-source-cards (assoc query :stages (get x :stages)))]
```

Comment:
what's the best way to do this upstream? ideally `pquery->field-usages` would receive a query that get all stages and joins resolved

Reply:
the downside is that calling this function will require a metaprovider to be bound.

cc @bshepherdson for ideas.

Reply:
yea, model shouldn't require QP ns.

> perhaps this code wants to run on the fully preprocessed query

this sounds nice. is there a helper like this, or should I roll one myself?